File: FBSimulatorControl/Strategies/FBProcessTerminationStrategy.m
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

#import "FBProcessTerminationStrategy.h"

#import <Cocoa/Cocoa.h>

#import <FBControlCore/FBControlCore.h>

#import "FBSimulatorError.h"

static const FBProcessTerminationStrategyConfiguration FBProcessTerminationStrategyConfigurationDefault = {
  .signo = SIGKILL,
  .options =
    FBProcessTerminationStrategyOptionsCheckProcessExistsBeforeSignal |
    FBProcessTerminationStrategyOptionsCheckDeathAfterSignal |
    FBProcessTerminationStrategyOptionsBackoffToSIGKILL,
};

@interface FBProcessTerminationStrategy ()

@property (nonatomic, strong, readonly) FBProcessFetcher *processFetcher;
@property (nonatomic, assign, readonly) FBProcessTerminationStrategyConfiguration configuration;
@property (nonatomic, strong, readonly) id<FBControlCoreLogger> logger;

@end

@interface FBProcessTerminationStrategy_WorkspaceQuit : FBProcessTerminationStrategy

@end

@implementation FBProcessTerminationStrategy_WorkspaceQuit

- (BOOL)killProcess:(FBProcessInfo *)process error:(NSError **)error
{
  // Obtain the NSRunningApplication for the given Application.
  NSRunningApplication *application = [self.processFetcher runningApplicationForProcess:process];
  // If the Application Handle doesn't exist, assume it isn't an Application and use good-ole kill(2)
  if ([application isKindOfClass:NSNull.class]) {
    [self.logger.debug logFormat:@"Application Handle for %@ does not exist, falling back to kill(2)", process.shortDescription];
    return [super killProcess:process error:error];
  }
  // Terminate and return if successful.
  if ([application terminate]) {
    [self.logger.debug logFormat:@"Terminated %@ with Application Termination", process.shortDescription];
    return YES;
  }
  // If the App is already terminated, everything is ok.
  if (application.isTerminated) {
    [self.logger.debug logFormat:@"Application %@ is Terminated", process.shortDescription];
    return YES;
  }
  // I find your lack of termination disturbing.
  if ([application forceTerminate]) {
    [self.logger.debug logFormat:@"Terminated %@ with Forced Application Termination", process.shortDescription];
    return YES;
  }
  // If the App is already terminated, everything is ok.
  if (application.isTerminated) {
    [self.logger.debug logFormat:@"Application %@ terminated after Forced Application Termination", process.shortDescription];
    return YES;
  }
  return [[[[FBSimulatorError
    describeFormat:@"Could not terminate Application %@", application]
    attachProcessInfoForIdentifier:process.processIdentifier processFetcher:self.processFetcher]
    logger:self.logger]
    failBool:error];
}

@end

@implementation FBProcessTerminationStrategy

+ (instancetype)withConfiguration:(FBProcessTerminationStrategyConfiguration)configuration processFetcher:(FBProcessFetcher *)processFetcher logger:(id<FBControlCoreLogger>)logger
{
  BOOL useWorkspaceKilling = (configuration.options & FBProcessTerminationStrategyOptionsUseNSRunningApplication) == FBProcessTerminationStrategyOptionsUseNSRunningApplication;
  return useWorkspaceKilling
    ? [[FBProcessTerminationStrategy_WorkspaceQuit alloc] initWithConfiguration:configuration processFetcher:processFetcher logger:logger]
    : [[FBProcessTerminationStrategy alloc] initWithConfiguration:configuration processFetcher:processFetcher logger:logger];
}

+ (instancetype)withProcessFetcher:(FBProcessFetcher *)processFetcher logger:(id<FBControlCoreLogger>)logger
{
  return [self withConfiguration:FBProcessTerminationStrategyConfigurationDefault processFetcher:processFetcher logger:logger];
}

- (instancetype)initWithConfiguration:(FBProcessTerminationStrategyConfiguration)configuration processFetcher:(FBProcessFetcher *)processFetcher logger:(id<FBControlCoreLogger>)logger
{
  NSParameterAssert(processFetcher);
  NSAssert(configuration.signo > 0 && configuration.signo < 32, @"Signal must be greater than 0 (SIGHUP) and less than 32 (SIGUSR2) was %d", configuration.signo);

  self = [super init];
  if (!self) {
    return nil;
  }

  _processFetcher = processFetcher;
  _configuration = configuration;
  _logger = logger;

  return self;
}

- (BOOL)killProcess:(FBProcessInfo *)process error:(NSError **)error
{
  BOOL checkExists = (self.configuration.options & FBProcessTerminationStrategyOptionsCheckProcessExistsBeforeSignal) == FBProcessTerminationStrategyOptionsCheckProcessExistsBeforeSignal;
  NSError *innerError = nil;
  if (checkExists && ![self.processFetcher processExists:process error:&innerError]) {
    return [[[[FBSimulatorError
      describeFormat:@"Could not find that process %@ exists", process]
      logger:self.logger]
      causedBy:innerError]
      failBool:error];
  }

  // Kill the process with kill(2).
  [self.logger.debug logFormat:@"Killing %@", process.shortDescription];
  if (kill(process.processIdentifier, self.configuration.signo) != 0) {
    // If the kill failed, then extract the error information and return.
    int errorCode = errno;
    if (errorCode == EPERM) {
      return [[[[FBSimulatorError
        describeFormat:@"Failed to kill process %@ as the sending process does not have the privelages", process.shortDescription]
        attachProcessInfoForIdentifier:process.processIdentifier processFetcher:self.processFetcher]
        logger:self.logger]
        failBool:error];
    }
    if (errorCode == ESRCH) {
      return [[[[FBSimulatorError
        describeFormat:@"Failed to kill process %@ as the sending process does not exist", process.shortDescription]
        attachProcessInfoForIdentifier:process.processIdentifier processFetcher:self.processFetcher]
        logger:self.logger]
        failBool:error];
    }
    if (errorCode == EINVAL) {
      return [[[[FBSimulatorError
        describeFormat:@"Failed to kill process %@ as the signal %d was not a valid signal number", process.shortDescription, self.configuration.signo]
        attachProcessInfoForIdentifier:process.processIdentifier processFetcher:self.processFetcher]
        logger:self.logger]
        failBool:error];
    }
    return [[FBSimulatorError
      describeFormat:@"Failed to kill process %@ with unknown errno %d", process.shortDescription, errorCode]
      failBool:error];
  }

  BOOL checkDeath = (self.configuration.options & FBProcessTerminationStrategyOptionsCheckDeathAfterSignal) == FBProcessTerminationStrategyOptionsCheckDeathAfterSignal;
  if (!checkDeath) {
    [self.logger.debug logFormat:@"Killed %@", process.shortDescription];
    return YES;
  }

  // It may take some time for the process to have truly died, so wait for it to be so.
  [self.logger.debug logFormat:@"Waiting on %@ to dissappear from the process table", process.shortDescription];
  if (![self.processFetcher waitForProcessToDie:process timeout:FBControlCoreGlobalConfiguration.fastTimeout]) {
    // If this is a SIGKILL and it's taken a while for the process to dissapear, perhaps the process isn't
    // well behaved when responding to other terminating signals.
    // There's nothing more than can be done with a SIGKILL.
    BOOL backoff = (self.configuration.options & FBProcessTerminationStrategyOptionsBackoffToSIGKILL) == FBProcessTerminationStrategyOptionsBackoffToSIGKILL;
    if (self.configuration.signo == SIGKILL || !backoff) {
      return [[[[FBSimulatorError
        describeFormat:@"Timed out waiting for %@ to dissapear from the process table", process.shortDescription]
        attachProcessInfoForIdentifier:process.processIdentifier processFetcher:self.processFetcher]
        logger:self.logger]
        failBool:error];
    }

    // Try with SIGKILL instead.
    FBProcessTerminationStrategyConfiguration configuration = self.configuration;
    configuration.signo = SIGKILL;
    [self.logger.debug logFormat:@"Backing off kill of %@ to SIGKILL", process.shortDescription];
    if (![[FBProcessTerminationStrategy withConfiguration:configuration processFetcher:self.processFetcher logger:self.logger] killProcess:process error:&innerError]) {
      return [[[[[FBSimulatorError
        describeFormat:@"Attempted to SIGKILL %@ after failed kill with signo %d", process.shortDescription, self.configuration.signo]
        attachProcessInfoForIdentifier:process.processIdentifier processFetcher:self.processFetcher]
        logger:self.logger]
        causedBy:innerError]
        failBool:error];
    }
  }

  [self.logger.debug logFormat:@"Killed %@", process.shortDescription];
  return YES;
}

- (BOOL)killProcesses:(NSArray<FBProcessInfo *> *)processes error:(NSError **)error
{
  for (FBProcessInfo *process in processes) {
    NSParameterAssert(process.processIdentifier > 1);
    if (![self killProcess:process error:error]) {
      return NO;
    }
  }
  return YES;
}

@end

File: FBControlCore/Processes/FBProcessTerminationStrategy.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

private let ProcessTableRemovalTimeout: TimeInterval = 20.0

/// An Option Set for Process Termination.
public struct FBProcessTerminationStrategyOptions: OptionSet, Sendable {
  public let rawValue: UInt
  public init(rawValue: UInt) {
    self.rawValue = rawValue
  }

  /// Checks for the process to exist before signalling.
  public static let checkProcessExistsBeforeSignal = FBProcessTerminationStrategyOptions(rawValue: 1 << 2)
  /// Waits for the process to die before returning.
  public static let checkDeathAfterSignal = FBProcessTerminationStrategyOptions(rawValue: 1 << 3)
  /// Whether to backoff to SIGKILL if a less severe signal fails.
  public static let backoffToSIGKILL = FBProcessTerminationStrategyOptions(rawValue: 1 << 4)
}

/// A Configuration for the Strategy.
public struct FBProcessTerminationStrategyConfiguration: Sendable {
  public var signo: Int32
  public var options: FBProcessTerminationStrategyOptions

  public init(signo: Int32, options: FBProcessTerminationStrategyOptions) {
    self.signo = signo
    self.options = options
  }
}

private let FBProcessTerminationStrategyConfigurationDefault = FBProcessTerminationStrategyConfiguration(
  signo: SIGKILL,
  options: [.checkProcessExistsBeforeSignal, .checkDeathAfterSignal, .backoffToSIGKILL]
)

public enum FBProcessTerminationStrategyError: Error, LocalizedError {
  case processDoesNotExist(processIdentifier: pid_t)
  case killFailed(processIdentifier: pid_t, message: String)
  case processTableRemovalTimedOut(processIdentifier: pid_t)
  case processDidNotDisappear(processIdentifier: pid_t, processInfo: String)
  case sigkillAfterFailedKill(processIdentifier: pid_t, signo: Int32, underlying: Error)

  public var errorDescription: String? {
    switch self {
    case let .processDoesNotExist(processIdentifier):
      return "Could not find that process \(processIdentifier) exists"
    case let .killFailed(processIdentifier, message):
      return "Failed to kill \(processIdentifier): '\(message)'"
    case let .processTableRemovalTimedOut(processIdentifier):
      return "Process \(processIdentifier) to be removed from the process table"
    case let .processDidNotDisappear(processIdentifier, _):
      return "Timed out waiting for \(processIdentifier) to disappear from the process table"
    case let .sigkillAfterFailedKill(processIdentifier, signo, _):
      return "Attempted to SIGKILL \(processIdentifier) after failed kill with signo \(signo)"
    }
  }
}

public final class FBProcessTerminationStrategy: NSObject {

  // MARK: Private Properties

  private let configuration: FBProcessTerminationStrategyConfiguration
  private let processFetcher: FBProcessFetcher
  private let workQueue: DispatchQueue
  private let logger: FBControlCoreLogger

  // MARK: Initializers

  public class func strategy(
    withConfiguration configuration: FBProcessTerminationStrategyConfiguration,
    processFetcher: FBProcessFetcher,
    workQueue: DispatchQueue,
    logger: FBControlCoreLogger
  ) -> Self {
    self.init(configuration: configuration, processFetcher: processFetcher, workQueue: workQueue, logger: logger)
  }

  public class func strategy(
    withProcessFetcher processFetcher: FBProcessFetcher,
    workQueue: DispatchQueue,
    logger: FBControlCoreLogger
  ) -> Self {
    self.init(
      configuration: FBProcessTerminationStrategyConfigurationDefault,
      processFetcher: processFetcher,
      workQueue: workQueue,
      logger: logger
    )
  }

  required init(
    configuration: FBProcessTerminationStrategyConfiguration,
    processFetcher: FBProcessFetcher,
    workQueue: DispatchQueue,
    logger: FBControlCoreLogger
  ) {
    precondition(
      configuration.signo > 0 && configuration.signo < 32,
      "Signal must be greater than 0 (SIGHUP) and less than 32 (SIGUSR2) was \(configuration.signo)")
    self.configuration = configuration
    self.processFetcher = processFetcher
    self.workQueue = workQueue
    self.logger = logger
    super.init()
  }

  // MARK: Public Methods

  @discardableResult
  public func killProcessIdentifier(_ processIdentifier: pid_t) -> FBFuture<NSNull> {
    let checkExists = hasOption(.checkProcessExistsBeforeSignal)
    if checkExists && processFetcher.processInfo(for: processIdentifier) == nil {
      return FBFuture(error: FBProcessTerminationStrategyError.processDoesNotExist(processIdentifier: processIdentifier))
    }

    logger.debug().log("Killing \(processIdentifier)")
    if kill(processIdentifier, configuration.signo) != 0 {
      return FBFuture(error: FBProcessTerminationStrategyError.killFailed(processIdentifier: processIdentifier, message: String(cString: strerror(errno))))
    }

    let checkDeath = hasOption(.checkDeathAfterSignal)
    if !checkDeath {
      logger.debug().log("Killed \(processIdentifier)")
      return FBFuture<NSNull>.empty()
    }

    logger.debug().log("Waiting on \(processIdentifier) to disappear from the process table")

    let waitFuture: FBFuture<NSNull> = waitForProcessIdentifierToDie(processIdentifier, on: workQueue, processFetcher: processFetcher)

    return
      waitFuture
      .onQueue(
        workQueue, timeout: ProcessTableRemovalTimeout,
        handler: { () -> FBFuture<AnyObject> in
          FBFuture<AnyObject>(error: FBProcessTerminationStrategyError.processTableRemovalTimedOut(processIdentifier: processIdentifier))
        }
      )
      .onQueue(
        workQueue,
        chain: { (future: FBFuture<AnyObject>) -> FBFuture<AnyObject> in
          if future.result != nil {
            self.logger.debug().log("Process \(processIdentifier) terminated")
            return FBFuture<NSNull>.empty().retyped(FBFuture<AnyObject>.self)
          }
          let backoff = self.hasOption(.backoffToSIGKILL)
          if self.configuration.signo == SIGKILL || !backoff {
            let processInfo: Any = self.processFetcher.processInfo(for: processIdentifier) ?? ("No Process Info" as NSString)
            return FBFuture(error: FBProcessTerminationStrategyError.processDidNotDisappear(processIdentifier: processIdentifier, processInfo: String(describing: processInfo)))
          }

          var newConfiguration = self.configuration
          newConfiguration.signo = SIGKILL
          self.logger.debug().log("Backing off kill of \(processIdentifier) to SIGKILL")
          let sigkillFuture: FBFuture<NSNull> = self.strategyWith(configuration: newConfiguration)
            .killProcessIdentifier(processIdentifier)

          return sigkillFuture.onQueue(
            self.workQueue,
            chain: { (innerFuture: FBFuture<AnyObject>) -> FBFuture<AnyObject> in
              if let error = innerFuture.error {
                return FBFuture(error: FBProcessTerminationStrategyError.sigkillAfterFailedKill(processIdentifier: processIdentifier, signo: self.configuration.signo, underlying: error))
              }
              return innerFuture
            })
        }
      ).retyped(FBFuture<NSNull>.self)
  }

  // MARK: Private

  private func hasOption(_ option: FBProcessTerminationStrategyOptions) -> Bool {
    configuration.options.contains(option)
  }

  private func strategyWith(configuration: FBProcessTerminationStrategyConfiguration) -> FBProcessTerminationStrategy {
    FBProcessTerminationStrategy(
      configuration: configuration,
      processFetcher: processFetcher,
      workQueue: workQueue,
      logger: logger
    )
  }

  private func waitForProcessIdentifierToDie(_ processIdentifier: pid_t, on queue: DispatchQueue, processFetcher: FBProcessFetcher) -> FBFuture<NSNull> {
    FBFuture<NSNull>.onQueue(
      queue,
      resolveWhen: {
        processFetcher.processInfo(for: processIdentifier) == nil
      })
  }
}
